synapse/lib/agenda.py

```diff
@@ -0,0 +1,614 @@
+            raise s_exc.NoSuchIden()
+
+        try:
+            heappos = self.apptheap.index(appt)
+        except ValueError:
+            pass  # this is OK, just a non-recurring appt that has no more records
+        else:
+            # If we're already the last item, just delete it
+            if heappos == len(self.apptheap) - 1:
+                del self.apptheap[heappos]
+            else:
+                # put the last item at the current position and reheap
+                self.apptheap[heappos] = self.apptheap.pop()
+                heapq.heapify(self.apptheap)
+
+        del self.appts[iden]
+        await self._hivedict.pop(s_common.ehex(iden))
+
+    async def _scheduleLoop(self):
+        '''
+        Task loop to issue query tasks at the right times.
+        '''
+        while True:
+            try:
+                timeout = None if not self.apptheap else self.apptheap[0].nexttime - time.time()
+                if timeout is None or timeout >= 0.0:
+                    await asyncio.wait_for(self._wake_event.wait(), timeout=timeout)
+            except asyncio.TimeoutError:
+                pass
+            if self.isfini:
+                return
+            self._wake_event.clear()
+
+            now = time.time()
+            while self.apptheap and self.apptheap[0].nexttime <= now:
+                appt = heapq.heappop(self.apptheap)
+                appt.updateNexttime(now)
+                if appt.nexttime:
+                    heapq.heappush(self.apptheap, appt)
+                if appt.isrunning:
+                    logger.warning(
+                        'Appointment %s is still running from previous time when scheduled to run.  Skipping.',
+                        s_common.ehex(appt.iden))
+                else:
+                    await self._execute(appt)
+
+    async def _execute(self, appt):
+        '''
+        Fire off the task to make the storm query
+        '''
+        if appt.username is None or self.core.auth is None:
+            user = None
+        else:
+            user = self.core.auth.users.get(appt.username)
+            if user is None:
+                logger.warning('Unknown username %s in stored appointment', appt.username)
+                return
+        await self.schedCoro(self._runJob(user, appt))
+
+    async def _runJob(self, user, appt):
+        '''
+        Actualy run the storm query, updating the appropriate statistics and results
+        '''
+        count = 0
+        appt.isrunning = True
+        appt.laststarttime = time.time()
+        appt.startcount += 1
+        await self._storeAppt(appt)
+        logger.info(f'Agenda executing as user {user}: query {appt.query}')
+        try:
+            async for _ in self.core.eval(appt.query, user=user):  # NOQA
+                count += 1
+        except asyncio.CancelledError:
+            result = 'cancelled'
+            raise
+        except Exception as e:
+            result = f'raised exception {e}'
```

Comment:
Do we want to log the trabeback here in the event that we encounter some weird error?

Reply:
testing this is difficult - could reach into the Agenda object itself and modify an Appt structure to have a bad query syntax.